hip-clang.docker

```diff
@@ -67,3 +67,15 @@ RUN export HIPCC_COMPILE_FLAGS_APPEND='-O3 -Wno-format-nonliteral -parallel-jobs
 # install last released miopentensile in default, install latest commits when MIOTENSILE_VER="latest"
 RUN if [ "$MIOTENSILE_VER" = "latest" ] ; then cget -p $PREFIX install ROCmSoftwarePlatform/MIOpenTensile@7568654c938d42e9a91c6b18fb382f5b978d12fd; else cget -p $PREFIX install ROCmSoftwarePlatform/MIOpenTensile@5fe0bf4a8dc59f3ab62df929297280915372ce16; fi
 
+RUN cd ~ && \
+    git clone -b miopen-dialect https://github.com/ROCmSoftwarePlatform/llvm-project-mlir.git && \
+    cd llvm-project-mlir && mkdir -p build && cd build && \
+    git reset --hard 5b26e1305d6ba82c385f560287f6af97057c7642 && \
+    cmake -G "Unix Makefiles" ../llvm \
+      -DLLVM_ENABLE_PROJECTS="mlir" \
+      -DCMAKE_BUILD_TYPE=Release \
+      -DBUILD_SHARED_LIBS=OFF \
+      -DLLVM_BUILD_LLVM_DYLIB=OFF \
+      -DLLVM_ENABLE_TERMINFO=OFF && \
+    make -j libMLIRMIOpen && \
```

Comment:
This should call `make package` and then do `dpkg -i <package-name>` so it can replace the llvmamdgpu package, otherwise we can end up with two different llvm versions.

Reply:
The only consumer of (this standalone static library) `libMLIRMIOpen.a` is MIOpen. It only has limited symbols that expose to MIOpen only. It expose zero llvm symbols.

Everything else in ROCm bit will continue to use `llvmamdgpu`.

By doing the above, our populated library stays independent from the ROCm shipped llvm. I do not see the two set of libraries can replace one another. Hope this clarifies. Let me know if

Reply:
> It only has limited symbols that expose to MIOpen only. It expose zero llvm symbols.

But wont this have internal llvm symbols? because when we are building rocm with `BUILD_SHARED_LIBS=Off` then we can't have symbols(even unexposed symbols) from two different llvm versions.

> By doing the above, our populated library stays independent from the ROCm shipped llvm.

Thats fine for prototyping and initial integration, but to be part of a rocm release this should be part of our llvm package.

Reply:
Wants to clarify/catch-up on this to make sure we are on the same page:

> ...building rocm with BUILD_SHARED_LIBS=Off...

I just tried to build with `CXX=/opt/rocm/llvm/bin/clang++ cmake -DMIOPEN_BACKEND=HIP -DBUILD_SHARED_LIBS=Off -DCMAKE_PREFIX_PATH="/opt/rocm/hcc;/opt/rocm/hip;/dockerx/deps" ..`. It ended up an linker error in `crc32`. I don't believe this linker issue is related with libMLIRMIOpen.

BTW, Does MIOpen have symbols from the ROCm shipped llvmamdgpu?

> ...we can't have symbols(even unexposed symbols)...
Why is that the case? Honestly I don't know all the linking details (to know how symbols are treated) but I did query `nm` on both libMIOpen.so and libMLIRMIOpen.a, no redundant llvm symbols showed up. 

According to a suggestion from @whchung, we used to conduct an experiment such that: we do dlopen() in our llvm application component on `libMIOpen.so` and things are working fine.

I'm thinking if you could suggest a way to test (or examine) such that this internal llvm symbols issue can show up? I'd be happily to give it a shot. Then we can start the conversation from there.

Reply:
> BTW, Does MIOpen have symbols from the ROCm shipped llvmamdgpu?

Indirectly, yes. MIOpen depends on HIP, which depends on comgr, which depends on llvm. So when HIP, comgr, and llvm are built with `-DBUILD_SHARED_LIBS=Off` then the user will need to link in the libs from llvmamdgpu.

Reply:
@pfultz2 Appreciate the explanation here. I just pinged @whchung on this. He offered me a way out:

We will populate the libMLIRMIOpen.a such that it strictly expose only the binary interface. All other (LLVM) symbols will be stripped away. Validate with `objdump -t`. This should prevent the issue from happening.

It is on my list of TODOs now. I want to scope this PR to non-COMGR. Then address this issue in MLIR. Finally file another PR to enable COMGR. Would you agree on this approach?

Reply:
Will address it in [this](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/80) issue. Marking as resolved now.